error.go

```diff
@@ -0,0 +1,69 @@
+package semver
+
+import (
+	"bytes"
+	"fmt"
+)
+
+var rangeErrs = [...]string{
+	"%s is less than the minimum of %s",
+	"%s is less than or equal to the minimum of %s",
+	"%s is greater than the maximum of %s",
+	"%s is greater than or equal to the maximum of %s",
+	"%s is specifically disallowed in %s",
+}
+
+const (
+	rerrLT = iota
+	rerrLTE
+	rerrGT
+	rerrGTE
+	rerrNE
+)
+
+type MatchFailure interface {
+	error
+	Pair() (v *Version, c Constraint)
+}
+
+type rangeMatchFailure struct {
+	v   *Version
+	rc  rangeConstraint
+	typ int8
+}
```

Comment:
Why are the error implementation private? If it were public an implementing application could do a type check to see which error actually happened if it wanted. Curious to your philosophy on error handling here.